### PR TITLE
EOS-20855:Mini-provisioner cleanup command fails if node is in standby

### PR DESCRIPTION
**Pull request by akash2144**
Signed-off-by: Akash Dudhane <akash.a.dudhane@seagate.com>

## Problem Statement
<pre>
  <code>
  Story Ref (if any): EOS-20855
    F-97D - Cluster: Factory Reset: Pre-Factory
    Your Problem statement here...
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes/No
  </code>
</pre>
## Problem Description
<pre>
  <code>
    1. F-97D - Cluster: Factory Reset: Pre-Factory
    2. Copy "cortx_ha_log.conf" file under " /etc/logrotate.d/" directory.
  </code>
</pre>
## Solution
<pre>
  <code> 
    system back to pre-factory stages when --pre-factory option is used with cleanup command
    Copied "cortx_ha_log.conf" file under " /etc/logrotate.d/" directory.
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Tested & confirmed that " /etc/logrotate.d/cortx_ha_log.conf" created after post_install command.
    and cleanup works fine with --pre-factory option.
  </code>
</pre>

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

**Comment by indrajitzagade** [comment]
@akash2144  Why the earlier patch #485 was closed?
Also, why the current PR is for upgrade branch?

File: ha/setup/ha_setup.py [hunk 1]
```diff
@@ -858,6 +862,18 @@ def process(self):
...
+            # Create a directory and copy config file
+            PostInstallCmd.copy_file(const.SOURCE_CONFIG_FILE, const.HA_CONFIG_FILE)
+            PostInstallCmd.copy_file(f"{const.SOURCE_CONFIG_PATH}/{const.CM_CONTROLLER_INDEX}.json",
+                            const.CM_CONTROLLER_SCHEMA)
+            PostInstallCmd.copy_file(const.SOURCE_ALERT_FILTER_RULES_FILE, const.ALERT_FILTER_RULES_FILE)
+            PostInstallCmd.copy_file(const.SOURCE_ALERT_EVENT_RULES_FILE, const.ALERT_EVENT_RULES_FILE)
+            PostInstallCmd.copy_file(const.SOURCE_CLI_SCHEMA_FILE, const.CLI_SCHEMA_FILE)
+            PostInstallCmd.copy_file(const.SOURCE_SERVICE_FILE, const.SYSTEM_SERVICE_FILE)
+            PostInstallCmd.copy_file(const.SOURCE_HEALTH_HIERARCHY_FILE, const.HEALTH_HIERARCHY_FILE)
+            PostInstallCmd.copy_file(const.SOURCE_IEM_SCHEMA_PATH, const.IEM_SCHEMA)
```
Comment (indrajitzagade) [inline ha/setup/ha_setup.py:876]: We need to copy the log rotate file too here, right?

**Review by indrajitzagade — CHANGES_REQUESTED** [review]
Please fix/reply review comments.

**Comment by akash2144** [comment]
> @akash2144 Why the earlier patch #485 was closed?
> Also, why the current PR is for upgrade branch?

@indrajitzagade  we discussed these in DSU, for testing reset & cleanup command we need (logrotate file) fix  in upgrade branch.

File: ha/setup/ha_setup.py (hunk 1 above)
Comment (akash2144) [reply ha/setup/ha_setup.py:876]: but we are not deleting logroatate file anywhere. These file remains there as it is.
@indrajitzagade If required plz let me know i will add logrotate file.

**Comment by indrajitzagade** [comment]
> 
> 
> > @akash2144 Why the earlier patch #485 was closed?
> > Also, why the current PR is for upgrade branch?
> 
> @indrajitzagade we discussed these in DSU, for testing reset (EOS-19786) & cleanup command we need (logrotate file) fix in upgrade branch.

Sure.

**Review by indrajitzagade — APPROVED** [review]

**Review by Madhura-08 — APPROVED** [review]

**Review by indrajitzagade — APPROVED** [review]

File: ha/setup/ha_setup.py [hunk 2]
```diff
@@ -858,6 +857,10 @@ def process(self):
                 self._confstore.delete(f"{const.PVTFQDN_TO_NODEID_KEY}/{node_name}")
             # Delete the config file
             self.remove_config_files()
+
+            Log.info("Post_install being called from cleanup command")
+            PostInstallCmd.process(self)
```
Comment (ajaysrivas) [inline ha/setup/ha_setup.py:862]: Is this tested? How can a member function be called like this?

Comment (akash2144) [reply ha/setup/ha_setup.py:862]: yes, Tested. it works fine
2021-07-07 07:16:19 ha_setup INFO [process] Post_install being called from cleanup command
2021-07-07 07:16:19 ha_setup INFO [process] Processing post_install command
2021-07-07 07:16:19 ha_setup INFO [process] cleanup: Copied HA configs file.
2021-07-07 07:16:19 ha_setup INFO [process] Found required cluster packages installed.
2021-07-07 07:16:30 ha_setup INFO [process] hacluster is a part of the haclient group
2021-07-07 07:16:30 ha_setup INFO [process] post_install command is successful
2021-07-07 07:16:30 ha_setup INFO [process] cleanup command is successful

Comment (ajaysrivas) [reply ha/setup/ha_setup.py:862]: It works but this is not the correct way instance method. process() method is not using any class attribute. This is the reason, there is no problem. If it start using those tomorrow, then the same code will break. Please make the changes.

File: ha/core/controllers/pcs/node_controller.py [hunk 3]
```diff
@@ -112,20 +112,24 @@ def standby(self, nodeid: str) -> dict:
...
         Log.info(f"Current {nodeid} status is {node_status}")
         if node_status == NODE_STATUSES.STANDBY.value:
-            status = f"Node {nodeid} is already running in standby mode."
-        elif node_status != NODE_STATUSES.ONLINE.value:
-            return {"status": const.STATUSES.FAILED.value, "output": "",
-                    "error": f"Failed to put node in standby as node is in {node_status}"}
-        else:
+            return {"status": const.STATUSES.SUCCEEDED.value, "output":
+                f"Node {nodeid} is already running in standby mode.", "error": ""}
+        elif node_status == NODE_STATUSES.STANDBY_WITH_RESOURCES_RUNNING.value:
```
Comment (indrajitzagade) [inline ha/core/controllers/pcs/node_controller.py:117]: There could be other node_status values that we need to handle here e.g. MAINTENANCE. Why are we changing this check?

Comment (indrajitzagade) [reply ha/core/controllers/pcs/node_controller.py:117]: Got it, we have else below.

File: ha/setup/ha_setup.py [hunk 4]
```diff
@@ -858,6 +861,10 @@ def process(self):
                 self._confstore.delete(f"{const.PVTFQDN_TO_NODEID_KEY}/{node_name}")
             # Delete the config file
             self.remove_config_files()
+            if CleanupCmd.PRE_FACTORY_CHECK is True:
```
Comment (indrajitzagade) [inline ha/setup/ha_setup.py:864]: I guess this check needs to be reverted:
If cleanup with pre-factory called: Revert things done in post_install command as well
else: do post_install after the current cleanup.

**Review by indrajitzagade — CHANGES_REQUESTED** [review]
Please take care of the review comments.

File: ha/setup/ha_setup.py (hunk 4 above)
Comment (akash2144) [reply ha/setup/ha_setup.py]: Yes, corrected

File: ha/const.py [hunk 5]
```diff
@@ -306,3 +308,6 @@ class CLUSTER_STATUS(Enum):
     DEGRADED = "degraded"
     UNHEALTHY = "unhealthy"
     ONLINE = "online"
+
+class MINI_PROVISION_CMD(Enum):
```
Comment (ajaysrivas) [inline ha/const.py:312]: rename it to MINI_PROVISIONER_PHASES.

File: ha/setup/ha_setup.py [hunk 6]
```diff
@@ -858,6 +861,10 @@ def process(self):
                 self._confstore.delete(f"{const.PVTFQDN_TO_NODEID_KEY}/{node_name}")
             # Delete the config file
             self.remove_config_files()
+            if CleanupCmd.PRE_FACTORY_CHECK is not True:
+                Log.info("Post_install being called from cleanup command")
+                self._execute.run_cmd(f"ha_setup {MINI_PROVISION_CMD.POST_INSTALL.value} --config {self._url}")
```
Comment (ajaysrivas) [inline ha/setup/ha_setup.py:866]: Create object instead of calling command.

Comment (akash2144) [reply ha/setup/ha_setup.py]: Done

**Review by indrajitzagade — APPROVED** [review]

**Pull request merged**